### PR TITLE
Update the SSH passthrough documentation

### LINKED ISSUE #18117: Using `exec -a` in `gitea/gitea:dev` container broke SSH passthrough access
### Gitea Version

Gitea version 1.16.0

### How are you running Gitea?

`gitea/gitea:dev` docker.

### Description

Gitea Docker now uses `exec -a "$0" ...` in `/usr/local/bin/gitea`, so AppPath is resolved to `/usr/local/bin/gitea` now.

Before, the AppPath is `/app/gitea/gitea`.

The AppPath is used inside authorized_keys file and shared on both container and host side.

This can introduce some problems:

1. The document for SSH pass-through are still using `/app/gitea/gitea`, which misleads users. https://docs.gitea.io/en-us/install-with-docker/#ssh-container-passthrough
2. It breaks existing Docker pass-through users.

The easy fix is using `exec ....` instead of `exec -a "$0" ...` (Is there any Pros of using `exec -a`?)

Or Gitea should update the document and mark this behavior as breaking.

by @wxiaoguang: personally I prefer to keep 1.16 behavior, otherwise it's too strange to force users creating a `/app/gitea/` on their host.

<details>

**I solved this issue (at least temporarily?) in the midst of writing a bug report. But it was caused by an (unknown) bug in `gitea` which may or may not be resolved. I'm going to leave this bug report here because the cause is unknown to me and it could happen again (e.g. when adding a key) and in case it helps others troubleshoot similar problems in the future.**

The short of it: some update to the `gitea/gitea:dev` container caused the `git` user's `authorized_keys` file to rewrite `/app/gitea/gitea` to `/usr/local/bin/gitea`, after I pulled the `gitea/gitea:dev` container for the first time in about a month or 2, and relaunched the image via `docker-compose.yml`. I am reasonably sure that the `gitea` 'database version' made a jump from `189` to `205` when I did this (maybe that is helpful information).

The issue
---

Last time I pulled the `gitea/gitea:dev` container, it was somewhat ahead of `1.15.8`. I pulled the `dev` version because it contained an unmerged bug fix regarding FIDO and 2FA.

Since pulling the `gitea/gitea:dev` container again yesterday (I forgot I had the `dev` tag set), I can no longer access the SSH passthrough, with this error:
```
local-pc $ git push
> bash: line 1: /usr/local/bin/gitea: No such file or directory
> fatal: Could not read from remote repository.
>
> Please make sure you have the correct access rights
> and the repository exists.
```

Troubleshooting and resolving
---
This error makes no sense to me, as I run on the remote server::
```
remote-server $ docker exec -it container_gitea bash
bash-5.1# ls -lahF /usr/local/bin
> total 10M    
> drwxr-xr-x    1 root     root        4.0K Dec 27 02:52 ./
> drwxr-xr-x    1 root     root        4.0K Dec 27 02:49 ../
> -rwxr-xr-x    1 root     root        9.4M Dec 27 02:52 environment-to-ini*
> -rwxr-xr-x    1 root     root         582 Dec 27 02:49 gitea*
```
Showing that `/usr/local/bin/gitea` is readable and executable by everyone.

Gitea is configured to run with a `git` user like so in the `docker-compose.yml`:
```yaml
  server:
    # Latest version as of now still includes a bug that doesn't allow FIDO enrolment unless 2FA is enabled.
    #image: gitea/gitea:1.15.8
    image: gitea/gitea:dev
    container_name: container_gitea
    environment:
      - USER_UID=${GIT_USER_UID?No GIT_USER_UID set in env file}
      - USER_GID=${GIT_USER_GID?No GIT_USER_GID set in env file}
      - GITEA__database__DB_TYPE=postgres
      - GITEA__database__HOST=db:5432
      - GITEA__database__NAME=gitea
      - GITEA__database__USER=gitea
      - GITEA__database__PASSWD=gitea
```

Naturally, I try downgrading to a previous version of `gitea`. I think that my old `gitea/gitea:dev` image before pulling was compatible with `1.15.8` (I left that comment you see in the `docker-compose.yml`). So I try reverting to that version. `gitea` fails to start however, and the `docker logs container_gitea` shows:
```
2021/12/29 12:42:17 ...les/storage/local.go:47:NewLocalStorage() [I] Creating new Local Storage at /data/gitea/repo-archive
2021/12/29 12:42:17 routers/init.go:93:GlobalInit() [I] SQLite3 Supported
2021/12/29 12:42:17 routers/common/db.go:20:InitDBEngine() [I] Beginning ORM engine initialization.
2021/12/29 12:42:17 routers/common/db.go:27:InitDBEngine() [I] ORM engine initialization attempt #1/10...
2021/12/29 12:42:17 ...om/urfave/cli/app.go:524:HandleAction() [I] PING DATABASE postgres
Downgrading database version from '205' to '189' is not supported and may result in loss of data integrity.
If you really know what you're doing, execute `UPDATE version SET version=189 WHERE id=1;`
2021/12/29 12:42:17 ...ations/migrations.go:414:Migrate() [F] Downgrading database version from '205' to '189' is not supported and may result in loss of data integrity.
        If you really know what you're doing, execute `UPDATE version SET version=189 WHERE id=1;`
Received signal 15; terminating.
```

Uh oh. I'm not messing around with a database downgrade, especially when the version difference looks so enormous. So I'm going to have to try to rectify this issue with the container I have. (I do `docker tag gitea/gitea:dev gitea/gitea:peg` and change the `docker-compose.yml` to use this. I don't want to be on `dev` anymore and when a new compatible fixed version tag becomes available, e.g. `gitea/gitea:1.16`, I will use that.)

Now I try simply `ssh`ing into the server to see what happens:
```
local-pc $ ssh git@remote-server
PTY allocation request failed on channel 0
bash: line 1: /usr/local/bin/gitea: No such file or directory
Connection to <hostname> closed.
```

So now I `docker stop` the `container_gitea` entirely, and I try `ssh`ing again and get the same result as above! Okay, so something wrong with the SSH passthrough configuration then?

I check and see that `/app/gitea/gitea` on `remote-server` is readable and executable by everything (as is `/app` and `/app/gitea`). Yes, that's fine.

Then I try `ssh`ing using a local user:
```
remote-pc $ ssh git@localhost
PTY allocation request failed on channel 0
bash: line 1: /usr/local/bin/gitea: No such file or directory
Connection to <hostname> closed.
```
Then with the `git` user which `gitea` runs as: **this works**.
```
git@remote-pc $ ssh git@localhost
> Last login: ....
git@remote-pc $ 
```

I don't understand. Here are the contents of `/app/gitea/gitea`:
```bash
#!/bin/sh
# Taken from https://docs.gitea.io/en-us/install-with-docker/#ssh-container-passthrough
ssh -p 2221 -o StrictHostKeyChecking=no git@127.0.0.1 "SSH_ORIGINAL_COMMAND=\"$SSH_ORIGINAL_COMMAND\" $0 $@"
```

So now I check `remote-server:/home/git/.ssh/authorized_keys`; this is where the magic is supposed to happen which forwards requests to `gitea`. Bingo. Here's the problem:
```bash
# gitea public key
command="/usr/local/bin/gitea --config=/data/gitea/conf/app.ini serv key-12",no-port-forwarding,no-X11-forwarding,no-agent-forwarding,no-pty,no-user-rc,restrict ssh-rsa <user1_SSH_PUB_KEY> user1@localhost
# gitea public key
command="/usr/local/bin/gitea --config=/data/gitea/conf/app.ini serv key-10",no-port-forwarding,no-X11-forwarding,no-agent-forwarding,no-pty,no-user-rc,restrict ssh-rsa <some_SSH_PUB_KEY>
# SSH pubkey from the Host machine's git user

ssh-rsa <gituser_SSH_PUB_KEY> Git Host Key

# Other keys from users of Gitea
command="/home/git/forward_ssh_to_gitea.sh --config=/data/gitea/conf/app.ini serv key-1",no-port-forward,no-X11-forwarding,no-agent-forwarding,no-pty <user pubkey>
```

I only ever added keys to gitea using the web interface, and I didn't do anything like `ssh-copy-id git@remote-server` to copy keys manually to that user. 

Thankfully gitea seems to keep backups of `authorized_keys`. I copied over one of the two most recent backups (same minute) sitting in the same directory next to `authorized_keys`. It looks like this:
```bash
# gitea public key
command="/app/gitea/gitea --config=/data/gitea/conf/app.ini serv key-10",no-port-forwarding,no-X11-forwarding,no-agent-forwarding,no-pty,no-user-rc,restrict ssh-rsa <some_SSH_PUB_KEY> 
# gitea public key
command="/app/gitea/gitea --config=/data/gitea/conf/app.ini serv key-12",no-port-forwarding,no-X11-forwarding,no-agent-forwarding,no-pty,no-user-rc,restrict ssh-rsa <user1_SSH_PUB_KEY> user1@localhost
# SSH pubkey from the Host machine's git user

ssh-rsa <gituser_SSH_PUB_KEY> Git Host Key

# Other keys from users of Gitea
command="/home/git/forward_ssh_to_gitea.sh --config=/data/gitea/conf/app.ini serv key-1",no-port-forward,no-X11-forwarding,no-agent-forwarding,no-pty <user pubkey>
```
After replacing `authorized_keys` with a copy of this backup, `ssh` and `git <command> git@remote-server` are now working.

</details>

### PR DESCRIPTION
**Pull request by zeripath**
Unfortunately #17846 broke the SSH documentation as the authorized_keys now uses
/usr/local/bin/gitea instead of /app/gitea/gitea.

This PR updates this documentation and provides documentation to offer multiple
options including docker-shells and ssh-shells to help avoid this problem in future.

Fix #18117

Signed-off-by: Andrew Thornton <art27@cantab.net>

### CODE REVIEW

**Review by techknowlogick — APPROVED** [review]

File: docs/content/doc/installation/with-docker.en-us.md [hunk 1]
```diff
@@ -307,77 +307,233 @@ To set required TOKEN and SECRET values, consider using Gitea's built-in [genera
...
+    EOF
+    chmod +x /usr/local/bin/gitea
+    ```
+
+  - For Gitea v1.15.x and earlier
+
+    ```bash
+    cat <<"EOF" | sudo tee /app/gitea/gitea
+    #!/bin/sh
+    ssh -p 2222 -o StrictHostKeyChecking=no git@127.0.0.1 "SSH_ORIGINAL_COMMAND=\"$SSH_ORIGINAL_COMMAND\" $0 $@"
```
Comment (wxiaoguang) [inline docs/content/doc/installation/with-docker.en-us.md:384]: Should the `$` (before SSH_ORIGINAL_COMMAND) be escaped?

And the `$0` and `$@` might to be quoted to `"$0"` and `"$@"`, as above For Gitea v1.16.0+ ( I am not sure, there are so many quoting levels  ...)

**Review by wxiaoguang — APPROVED** [review]

File: docs/content/doc/installation/with-docker.en-us.md (hunk 1 above)
Comment (zeripath) [reply docs/content/doc/installation/with-docker.en-us.md:384]: The "EOF" heredoc means the $ gets put in the script as seen meaning we don't need to quote the $ and the $ expansion is occuring at the right place now. 

So the only question is whether we need to double quote to deal with whitespace. 

We don't need "$0" as the suggested file names don't include spaces but we could add.

The $@ shouldn't need double quoted as there are no places that a space should be in the argument. I'm not completely sure of the failure modes though as this is reporting as working without the double quoting as it stands and certainly works in my testing.

The problem with double quoting the "$@" is whether the this results in: "$1" "$2" "$3" ... or whether this results in "$1 $2 $3 ...". Normally "$@" should result in effectively the first expansion but it's already within a double quoted string so I don't know what will happen. 

I am suspicious though that double quoting might actually only work in some configs and not others but I dunno.

Have you been able to test and show that double quoting works here?

Comment (wxiaoguang) [reply docs/content/doc/installation/with-docker.en-us.md:384]: For question 1 (the `$`), it's different for line 374 (`...=\"\$SSH_...`) and line 384 (`...=\"$SSH_...`) , one is escaped while another isn't.

~~For question 1 and 2: I think it should be `ssh {options} SSH_ORIGINAL_COMMAND="$SSH_ORIGINAL_COMMAND" "$0" "$@"`~~

Sorry, I misread about question 2, it's really complex, let me see ...

Comment (wxiaoguang) [reply docs/content/doc/installation/with-docker.en-us.md:384]: Well, for question 2, it's so tricky to do correct quoting with bash (I just found this: https://unix.stackexchange.com/questions/518012 ) , so maybe we had better leave it as it was.

File: docs/content/doc/installation/with-docker.en-us.md [hunk 2]
```diff
@@ -307,77 +307,233 @@ To set required TOKEN and SECRET values, consider using Gitea's built-in [genera
...
+  ```
+
+- The next step is to create the file that will issue the SSH forwarding from the host to the container. The name of this file depends on your version of Gitea:
+
+  - For Gitea v1.16.0+:
+
+    ```bash
+    cat <<"EOF" | sudo tee /usr/local/bin/gitea
+    #!/bin/sh
+    ssh -p 2222 -o StrictHostKeyChecking=no git@127.0.0.1 "SSH_ORIGINAL_COMMAND=\"\$SSH_ORIGINAL_COMMAND\" $0 $@"
```
Comment (zeripath) [inline docs/content/doc/installation/with-docker.en-us.md:374]:
```suggestion
    ssh -p 2222 -o StrictHostKeyChecking=no git@127.0.0.1 "SSH_ORIGINAL_COMMAND=\"$SSH_ORIGINAL_COMMAND\" $0 $@"
```

File: docs/content/doc/installation/with-docker.en-us.md (hunk 1 above)
Comment (zeripath) [reply docs/content/doc/installation/with-docker.en-us.md:384]: ah sorry I think the `\$` should be `$` !

File: docs/content/doc/installation/with-docker.en-us.md (hunk 2 above)
Comment (zeripath) [inline docs/content/doc/installation/with-docker.en-us.md]:
```suggestion
    ssh -p 2222 -o StrictHostKeyChecking=no git@127.0.0.1 "SSH_ORIGINAL_COMMAND=\"$SSH_ORIGINAL_COMMAND\" $0 $@"
```

File: docs/content/doc/installation/with-docker.en-us.md [hunk 3]
```diff
@@ -307,77 +307,233 @@ To set required TOKEN and SECRET values, consider using Gitea's built-in [genera
...
+The idea of this option is to use (essentially unchanged) the authorized_keys that gitea creates on the docker and use a special shell for git user that uses ssh to shell to the docker git user.
+
+- In this case we setup as above except instead of creating `/usr/local/bin/gitea` or `/app/gitea/gitea`
+we create a new shell for the git user:
+
+  ```bash
+  cat <<"EOF" | sudo tee /home/git/ssh-shell
+  #!/bin/sh
+  shift
+  ssh -p 2222 -o StrictHostKeyChecking=no git@127.0.0.1 "SSH_ORIGINAL_COMMAND=\"\$SSH_ORIGINAL_COMMAND\" $@"
```
Comment (zeripath) [inline docs/content/doc/installation/with-docker.en-us.md]:
```suggestion
  ssh -p 2222 -o StrictHostKeyChecking=no git@127.0.0.1 "SSH_ORIGINAL_COMMAND=\"$SSH_ORIGINAL_COMMAND\" $@"
```

**Review by zeripath — COMMENTED** [review]
remove spurious \

**Pull request merged**